carbonmark-api/src/utils/helpers/fetchMarketplaceListings.ts

```diff
@@ -96,21 +121,64 @@ export const fetchMarketplaceListings = async (
       },
     };
   });
+  return listingsWithProfiles;
+};
 
-  const activitiesWithProfiles = formattedActivities.map((act) => {
-    const activityWithHandles = { ...act };
-    const sellerData = profilesMap.get(act.seller.id.toLowerCase());
-    if (sellerData) {
-      set(activityWithHandles, "seller.handle", sellerData.handle);
-    }
-    if (act.buyer) {
-      const buyerData = profilesMap.get(act.buyer.id.toLowerCase());
-      if (buyerData && buyerData.handle) {
-        set(activityWithHandles, "buyer.handle", buyerData.handle);
-      }
-    }
-    return activityWithHandles;
+/**
+ * Query the subgraph for active marketplace listings and project data for the given project
+ * Filters out deleted, sold-out and inactive listings
+ * Fetches seller profile info from firebase
+ */
+export const fetchMarketplaceListings = async (
+  sdk: GQL_SDK,
+  { key, vintage, expiresAfter, fastify }: ListingsParams
+): Promise<[Listing[], Activity[]]> => {
+  const { project } = await sdk.marketplace.getProjectById({
+    projectId: key + "-" + vintage,
+    expiresAfter: expiresAfter ?? String(Math.floor(Date.now() / 1000)),
   });
+  const filteredListings = project?.listings?.filter(isActiveListing) || [];
+  const filteredActivities =
+    project?.activities?.filter(filterUnsoldActivity) || [];
+
+  const activitiesWithProfiles = await formatActivities(
+    filteredActivities,
+    fastify
+  );
+  const listingsWithProfiles = await formatListings(filteredListings, fastify);
 
   return [listingsWithProfiles, activitiesWithProfiles];
 };
+
+/**
+ * Query the subgraph for activities for the given project
+ */
+export const fetchProjectActivities = async (
+  sdk: GQL_SDK,
+  { projectId, activityType, fastify }: ActivitiesParams
+): Promise<Activity[]> => {
+  let data;
+  // FIXME: Is it possible to do that with the same gql query template?
+  // I suppose we could generate the gql programmatically but I think we do not want to bypass the templating system
```

Comment:
Could you expand on what you mean here?

Reply:
What I would like to do is this: 
```
   data = await sdk.marketplace.getAllActivities({
      projectId: formattedProjectId,
      activityType,
    });
```

And I would like the gql query used by the `sdk.marketplace.getAllActivities`  function to be smart enough to not filter on projects if `formattedProjectId` is an empty list.

Basically have the `if (projectId.length) {` logic made at the gql query level.